lib/WebService/Async/Segment.pm

```diff
@@ -0,0 +1,282 @@
+        group_id   => 'groupId',
+        device     => {
+            advertising_id      => 'advertisingId',
+            ad_tracking_enabled => 'adTrackingEnabled'
+        },
+    },
+};
+
+our $VERSION = '0.001';
+
+=head1 NAME
+
+WebService::Async::Segment - Unofficial support for the Segment service
+
+=head1 DESCRIPTION
+
+This class acts as a L<Future>-based async Perl wrapper for segment HTTP API.
+
+=cut
+
+=head1 METHODS
+
+=head2 configure
+
+Overrides the same method of the parent class L<IO::Async::Notifier>; required for object initialization.
+
+parameters:
+
+=over 4
+
+=item * C<write_key> - the API token of a Segment source.
+
+=item * C<base_uri> - the base uri of the Segment host, primarily useful for setting up test mock servers.
+
+=back
+
+=cut
+
+sub configure {
+    my ($self, %args) = @_;
+
+    for my $k (qw(write_key base_uri)) {
+        $self->{$k} = delete $args{$k} if exists $args{$k};
+    }
+
+    $self->next::method(%args);
+}
+
+=head2 write_key
+
+API token of the intended Segment source
+
+=cut
+
+sub write_key { shift->{write_key} }
+
+=head2 base_uri
+
+Server endpoint. Defaults to C<< https://api.segment.io/v1/ >>.
+
+Returns a L<URI> instance.
+
+=cut
+
+sub base_uri {
+    my $self = shift;
+    return $self->{base_uri} if blessed($self->{base_uri});
+    $self->{base_uri} = URI->new($self->{base_uri} // SEGMENT_BASE_URL);
+    return $self->{base_uri};
+}
+
+=head2 ua
+
+A L<Net::Async::HTTP> object acting as HTTP user agent
+
+=cut
+
+sub ua {
+    my ($self) = @_;
+
+    return $self->{ua} if $self->{ua};
+
+    $self->{ua} = Net::Async::HTTP->new(
+        fail_on_error            => 1,
+        decode_content           => 1,
+        pipeline                 => 0,
+        stall_timeout            => TIMEOUT,
+        max_connections_per_host => 2,
+        user_agent               => 'Mozilla/4.0 (WebService::Async::Segment; BINARY@cpan.org; https://metacpan.org/pod/WebService::Async::Segment)',
+    );
+
+    $self->add_child($self->{ua});
+
+    return $self->{ua};
+}
+
+=head2 basic_authentication
+
+Settings required for basic HTTP authentication
+
+=cut
+
+sub basic_authentication {
+    my $self = shift;
+
+    #C<Net::Async::Http> basic authentication information
+    return {
+        user => $self->write_key // '',
+        pass => ''
+    };
+}
+
+=head2 method_call
+
+Makes a Segment method call. It automatically defaults C<sent_at> to the current time and C<< context->{library} >> to the current module.
+
+It takes the following named parameters:
+
+=over 4
+
+=item * C<method> - required. Segment method name (such as B<identify> and B<track>).
+
+=item * C<args> - optional. Method arguments represented as a dictionary. This may include either common, method-specific or custom fields.
+
+=back
+
+Please refer to L<https://segment.com/docs/spec/common/> for a full list of common fieds supported by Segment.
+
+It returns a L<Future> object that should be taken care of by the caller.
```

Comment:
This still isn't very clear - "taken care of" doesn't really explain "caller owns the object, the request may be cancelled if you let it drop out of scope".

Generally you'd be better off having ownership in this class, especially given the expected fire-and-forget nature of this module.

Reply:
@tom-binary I didn't get it. Which `ownership` do you mean?

Reply:
Do you mean the ownership of `Customer`s?

Reply:
Do you mean simply calling `POST` without chaining?

Reply:
Adding a `retain` at the end of the Future chain, this code works fine:
```
sub experiment{
    my $customer = $segment->new_customer(%$customer_info);
    $customer->identify();
    $customer->track(event => 'test');
}
experiment();
$loop->delay_future(after => 1)->get;
```